### PR TITLE
feat: Reworked binary_search.cpp

### PR DESCRIPTION
**Pull request by Lazeeez**
#### Description of Change

The original binary_search.cpp was pretty blatant with no documentation, did some rework on it and added some testcases to check the algorithm.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: Did some rework on the original binary_search.cpp

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1854"></a>

### CODE REVIEW

**Comment by Lazeeez** [comment]
@Panquesito7 Can you please review this PR.

File: search/binary_search.cpp [hunk 1]
```diff
@@ -1,56 +1,148 @@
...
- * \returns index if T is found
- * \return -1 if T is not found
- */
-int binary_search(int a[], int r, int key) {
-    int l = 0;
-
-    while (l <= r) {
-        int m = l + (r - l) / 2;
-        if (key == a[m])
+ *algorithm](https://en.wikipedia.org/wiki/Binary_search_algorithm)
```
Comment (Panquesito7) [inline search/binary_search.cpp:4]:
```suggestion
 * algorithm](https://en.wikipedia.org/wiki/Binary_search_algorithm)
```

File: search/binary_search.cpp [hunk 2]
```diff
@@ -1,56 +1,148 @@
...
+ * continues in the upper half of the array. By doing this, the algorithm
+ * eliminates the half in which the target value cannot lie in each iteration.
+ *
+ * ### Complexities
+ *
+ * //n is the number of element in the array.
+ *
+ * Worst-case time complexity	O(log n)
+ * Best-case time complexity	O(1)
+ * Average time complexity	    O(log n)
```
Comment (Panquesito7) [inline search/binary_search.cpp:30]:
```suggestion
 * Average time complexity      O(log n)
```

File: search/binary_search.cpp [hunk 3]
```diff
@@ -1,56 +1,148 @@
...
+ * @namespace binary_search
+ * @brief Binary search searching algorihm
+ *******************************************************************************/
+namespace binary_search {
+
+/******************************************************************************
+ * @brief The main function which implements binary search
+ * @param arr vector to be searched in
+ * @param val value to be searched
+ * @returns @param int index of val in vector arr
```
Comment (Panquesito7) [inline search/binary_search.cpp:57]:
```suggestion
 * @returns integer index of `val` in vector `arr`
```

File: search/binary_search.cpp [hunk 4]
```diff
@@ -1,56 +1,148 @@
...
 
-    // this loop use for store value in Array
-    for (int i = 0; i < n; i++) {
-        std::cin >> a[i];
-    }
+/*******************************************************************************
+ * @brief Self-test implementation #1
+ * @returns void
+ *******************************************************************************/
+static void test1() {
```
Comment (Panquesito7) [inline search/binary_search.cpp:90]: Let's make this only `void` and create a static `tests` function, which will run all tests. 😄
```suggestion
void test1() {
```

File: search/binary_search.cpp [hunk 5]
```diff
@@ -1,56 +1,148 @@
...
+    // array = [1,31,231,12,12,2,5,51,21,23,12,3] , Value = 5
+    // should return 8
+    std::vector<uint64_t> arr = {{1, 31, 231, 12, 2, 5, 51, 21, 23, 12, 3}};
+    std::sort(arr.begin(), arr.end());
+    uint64_t expected_ans = 8;
+    uint64_t derived_ans = search::binary_search::binarySearch(arr, 31);
+    std::cout << "Test #3: ";
+    assert(derived_ans == expected_ans);
+    std::cout << "Passed!" << std::endl;
+}
```
Comment (Panquesito7) [inline search/binary_search.cpp:136]:
```suggestion
}
/*******************************************************************************
 * @brief Self-test implementations
 * @returns void
 *******************************************************************************/
static void tests() {
    test1();
    test2();
    test3();
}
```

File: search/binary_search.cpp [hunk 6]
```diff
@@ -1,56 +1,148 @@
...
-    else
-        std::cout << key << " not found" << std::endl;
+/*******************************************************************************
+ * @brief Main function
+ * @returns 0 on exit
+ *******************************************************************************/
+int main() {
+    test1();  // run self-test implementation #1
+    test2();  // run self-test implementation #2
+    test3();  // run self-test implementation #3
```
Comment (Panquesito7) [inline search/binary_search.cpp:145]:
```suggestion
    tests();  // run self-test implementations
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 😄👍

**Comment by Lazeeez** [comment]
@Panquesito7 Thanks for the suggestions and corrections. 😄

**Review by Panquesito7 — APPROVED** [review]
LGTM 🚀 Thank you for your contribution! 😄👍

**Review by Qannaf — APPROVED** [review]

**Pull request merged**